### ISSUE TITLE
Possible improvements: control number of errors and exception raised

### ISSUE THREAD

**Issue by gpetretto** [materialsproject/custodian#87]
Dear all, I opened this issue not to signal a problem, but rather to propose a couple of improvements to the current custodian implementation. These ideas emerged since we would like to use custodian as a base for error handling in a new project and we think that these functionalities would help us in our use cases.

#### Number of errors
The first modification is about the introduction of a finer control on the number errors before stopping the job. In particular I think that it would be nice to have the possibility to decide the maximum number of times a correction from a single error handler should be applied. 

I can imagine a series of cases where this could be useful:
* a correction for an error message that is only meaningful if applied once. Applying again the same correction could be useless, so stop the job immediately.
* in the case of the restart, one may want to control exactly how many times a calculation should be restarted before giving an error. This could be handled with the total number of errors, but in general it is not possible to predict how many other errors could be present.
* the VASP `CheckpointHandler` suggests that this handler should be used alone and with a high value for `max_errors`. However I suppose that the limitation of being the single error handler could be lifted if it is possible to control the maximum number of tolerated corrections for each of the other handlers.

Of course something like this could still be implemented at the level of each single error handler, but this will likely lead to the writing of a lot of redundant code, while it seems a sufficiently general feature to be present in the core implementation.

I have prepared a rough implementation of this modification here: https://github.com/gpetretto/custodian/commit/3ff38fce407adbaa8607c22abb992665be9c9c26

Such a modification should have no impact on existing code, as by default everything should still keep working as before. The only possible problem could be if someone has implemented some hanlders outside the official custodian implementation with names clashing with the new ones introduced here, but this seems a relatively remote possibility.

Here is a snippet of code demonstrating possible use cases: https://gist.github.com/gpetretto/b14a1addf53a8e99893fca24228e1d4e

#### More informative exceptions

I suppose that more often than not Custodian runs inside some other python code (e.g. fireworks), but, as far as I can see, what happens inside custodian is not really transparent from the outer layers. To give a more specific example, we would like to take advantage of the exact final state of the error handler/validator that caused the program to stop, to take further actions. This may involve a dynamical evolution of the workflow, like adding an intermediate step calculation before rerunning the current job. 
I would like to stress that here I am *not* suggesting to introduce some kind of integration with other python packages (being it fireworks or anything else), but just to let out more information when errors happen for whoever may be interest in having it. 

It seems that currently some information could be extracted from the `Custodian` instance (from the `log_run`, `total_errors` and `errors_current_job`), but this implies a bit of investigation, and maybe of guessing about what really happened inside custodian (e.g. is there any way one could know which validator made the calculation fail, aside from parsing the text of the error message?). 
Here is one possible modification of the current behavior that would already allow to directly receive some useful information
https://github.com/gpetretto/custodian/commit/95975957b06ac8ab491cfc5db874c13eea3b1cfb
https://gist.github.com/gpetretto/13f8f00994d62844cadeae812559bfb9
This takes advantage of the `validator` attribute in `CustodianError`, that is set, but it seems to be never used. Also, since the new error class `CustodianRuntimeError` subclasses `RuntimeError`, this should preserve almost entirely the current behavior, with the only difference being the name of the exception in the message. In addition it will provide additional information about what happened inside custodian.

However, if a bit more freedom could be allowed, I would prefer introducing a more detailed hierarchy of `CustodianError` errors and let these exception directly emerge from Custodian, so that one could be able to catch and analyze more in details each case differently. For example something like this:

```
RuntimeError
 +-- CustodianError
      +-- ValidationError
      +-- HandlerError
      +-- ReturnCodeError
      +-- MaxErrorsError
      +-- MaxErrorsPerJobError
```

Notice that I suggest to keep it a subclass of `RuntimeError` to ensure that any existing code that relies on catching `RuntimeError` will keep working.

Here is a possible implementation:
https://github.com/gpetretto/custodian/commit/d2e01054160e4d45a09101a2536616cff7eeb0e1
https://gist.github.com/gpetretto/45506abee33f68b103639399b2c77064

I would be interested in hearing comments and knowing if these kind of modifications could be of interest for you. If that's the case I can clean up the implementations and open pull requests.

Thanks

**Comment by computron** [comment]
Hi Guido,

I admittedly only skimmed most of this issue but I wanted to point out two things that seemed missing from the description:

1. Many of the error handlers already keep track of how many times they are run, and change strategy or quit depending on this. See for example the handling of brmix errors:
custodian/vasp/handlers.py:163

There is an internal count of how many times the BRMIX fix has been applied, and we try different strategies depending on the count. A similar strategy would be used to make sure an error fix is applied only once. it just requires setting up an error_count Counter like VaspErrorHandler does.

2. For the "More informative exceptions", I wasn't sure why you didn't mention the current "custodian.json" file that's written out.

**Comment by gpetretto** [comment]
Thanks for the comments.

For the first point this is also more or less how I have started to implement the handlers, but still it seemed that having this in the Custodian base class and in the check process could help speeding up the writing and make them cleaner.

Concerning the second point, it seems that the `custodian.json` contains the same information as the `log_run` attribute of the `Custodian` instance that I mentioned and to which one can have direct access. Still, correct me if I am wrong, this just contains the actions and strings to describe the error, but will not give direct access to the actual state of the handler instance. Also, it seems that no information about the validation is stored there and even when the error is caused by the reaching of the maximum number of errors.

**Comment by shyuep** [comment]
I am not sure what's the advantage of keeping track it in the Custodian base class. Keeping track of the state in the handler and validator makes more sense to me. Even if we want to do this, I would prefer the Custodian class to generically keep track of all information by having the handler return information during check().

As for expanding custodian.json to log more info, this is perfectly fine with me.

**Comment by gpetretto** [comment]
Thanks for the reply and sorry, I think that while trying to summarize I was instead sloppy in the last post. I actually intended to put the tracking of the number of corrections in `ErrorHandler`. See e.g. https://github.com/gpetretto/custodian/commit/3ff38fce407adbaa8607c22abb992665be9c9c26

As for having more information in case of errors, do I understand correctly that instead of having different kind of exceptions, as in the original suggestion, you would prefer to increase the amount of information in `log_run` (and  then `custodian.json`)? 

If that's the case, presently the `log_run` contains a list of dict of this kind: `{"job": job.as_dict(), "corrections": []}`. Since I imagine you would like to keep this intact, should I aim to handle it with something like this:
```python
{"job": job.as_dict(), # this could also be replaced by just job, as it will be serialized in dumpfn
"corrections": []
"handler": None,   # the instance of the handler that made the job fail correction, i.e. no correction was applied
"validator": None, # if the validation fails, set here the validator that failed
"max_errors": False, # set to true if the maximum number of errors has been reached
"max_errors_per_job": False # set to true if the maximum number of errors per job has been reached
}
```
Do you have any suggestion about how to improve this?
I would probably still prefer to have more information at the level of the exception raised, but increasing the `log_run` should be fine as well. In principle both could be added, serving different purposes: one for better integration with third-party modules and the other to have a more detailed logging.

**Comment by shyuep** [comment]
What you suggested is fine. Thanks.
